### DISCUSSION TITLE
Parametric generators: Was "warning: Macro %1 defined but not used within scope" removed in RPM 4.17+?

### DISCUSSION THREAD

**Discussion by hroncok** [rpm-software-management/rpm#2501] [Q&A] [answered]
Hello.

We have a workaround in our Fedora package: https://src.fedoraproject.org/rpms/python-rpm-generators/blob/b1fa63bf02b03b2120d32eb91ca2911d4ec77beb/f/pythonname.attr#_6 (see the comment above this line)

While debugging a related issue on CentOS Stream 9, I figured out that the warning is no longer present since Fedora 35.

### reproducer.spec

```
Name:           reproducer
Version:        0
Release:        0
Summary:        ...
License:        MIT
BuildArch:      noarch

%description
...

%prep

%build

%install
mkdir -p %{buildroot}
touch %{buildroot}/xxx

%files
/xxx
```

### /usr/lib/rpm/fileattrs/xxx.attr

```
%__xxx_provides() xxx
%__xxx_path ^/
```

### rpm-4.16.1.3-1.fc34

```
$ rpmbuild -ba reproducer.spec
...
Processing files: reproducer-0-0.noarch
warning: Macro %1 defined but not used within scope
Provides: reproducer = 0-0 xxx
...
```

### rpm-4.17.1-3.fc35

```
$ rpmbuild -ba reproducer.spec
...
Processing files: reproducer-0-0.noarch
Provides: reproducer = 0-0 xxx
...
```

Was this warning removed on purpose and can we safely remove the ugly workaround? Or is this a regression?

Looking through the relevant code's history near https://github.com/rpm-software-management/rpm/blob/a97c376f379ee48c3b0e92e9c79f64c9fd954b64/rpmio/macro.c#L946 does not reveal any obvious intended change.

**Comment by pmatilai** [comment, marked as answer]
Oh, sorry. I remember seeing this but it's fallen through the cracks.

The warning was an artifact of a hack in the rpm level, fixed in commit a9bae82f699a9b112c8802eebe5da9fff5a8f34b (in >= 4.17).  Rpm never warns on unused automatic macros such as %1, but in 4.16 the %1 passed to the parametric macro was not an automatic macro but a hardcoded hack. 

That said, a generator not using the filename passed to it via %1 sounds suspicious in itself.

**Reply by hroncok** [reply]
> That said, a generator not using the filename passed to it via %1 sounds suspicious in itself.

It's generating a provide based on the package name.
